### ISSUE TITLE
Tracking issue for naked fns (RFC #1201)

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#32408]
This is the tracking issue for rust-lang/rfcs#1201.

**Note that this is intended as a (particularly) experimental feature.** 

Feedback on its design is required, as well as comparisons with [global asm](https://github.com/rust-lang/rust/issues/35119).

**Comment by jackpot51** [comment]
This issue was partially addressed by https://github.com/rust-lang/rust/pull/29189/

**Comment by jackpot51** [comment]
https://github.com/rust-lang/rust/pull/32410 is an up to date implementation

**Comment by ranma42** [comment]
It might be a little late to point it out, but the RFC does not say anything about the interaction between naked functions and unwinding/panics.

**Comment by nikomatsakis** [comment]
Note that [global-asm](https://github.com/rust-lang/rust/issues/35119) has also been added, which overlaps somewhat with naked fns.

**Comment by mark-i-m** [comment]
So how exactly do experimental features get stabilized? This is an incredibly useful feature for building an OS, and I would love to see it move forward :)

**Comment by nikomatsakis** [comment]
@mark-i-m well, if we retrofit this into the "experimental RFC" process that we are experimenting with (so much experimenting!), then the next step would be to take the lessons we've learned from having the feature around and write-up a new RFC that lays out the:

- final design of the feature, and in particular the various tricky cases and how we decided to resolve them
- talks about experience in the wild
- etc

and then we would proceed as normal.

In the case of naked functions, I think if anything my doubts have grown *more* grave, not less. In particular, I still don't think we can say much with confidence about how Rust code embedded in naked functions is expected to be compiled. Right now it's all pretty darn unspecified, iirc. I'd be interested to see some kind of data points for whether all naked fns in the wild use purely inline assembly, or whether they embed Rust code, and if so, *what* Rust code they embed (maybe we can define some kind of sensible subset)?

**Comment by mark-i-m** [comment]
@nikomatsakis Here is some code I wrote last weekend for an OS kernel: https://gist.github.com/mark-i-m/361cbcc39769f965b1c419091b9cbf4f

It is mostly assembly, but it does have Rust mixed in. Specifically, it has function calls not just to other naked or inlined functions, but also to normal Rust functions, such as [here](https://gist.github.com/mark-i-m/361cbcc39769f965b1c419091b9cbf4f#file-machine-rs-L348).

> In particular, I still don't think we can say much with confidence about how Rust code embedded in naked functions is expected to be compiled.

Personally, my expectation is that it should compile just like it normally would -- just the prologue/epilogue are left out. For example,

```rust
extern "C" fn foo(bar: usize) -> usize {
    bar << 1
}
```
generates
```asm
_ZN10playground3foo17h002f6b2acb4d5b2dE:
.Lfunc_begin2:
	.loc	1 2 0
	.cfi_startproc
	pushq	%rbp
.Ltmp12:
	.cfi_def_cfa_offset 16
.Ltmp13:
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
.Ltmp14:
	.cfi_def_cfa_register %rbp
	movq	%rdi, -16(%rbp)
.Ltmp15:
	.loc	1 2 0 prologue_end
	movq	-16(%rbp), %rdi
	movq	%rdi, -8(%rbp)
.Ltmp16:
	.loc	1 3 0
	movq	-8(%rbp), %rdi
	shlq	$1, %rdi
	movq	%rdi, -24(%rbp)
	.loc	1 4 0
	movq	-24(%rbp), %rax
	popq	%rbp
	retq
```

I imagine the `naked` version would simply be this (I didn't adjust the offsets on the stack):

```asm
_ZN10playground3foo17h002f6b2acb4d5b2dE:
.Ltmp15:
	.loc	1 2 0 prologue_end
	movq	-16(%rbp), %rdi
	movq	%rdi, -8(%rbp)
.Ltmp16:
	.loc	1 3 0
	movq	-8(%rbp), %rdi
	shlq	$1, %rdi

```

**Comment by mark-i-m** [comment]
Oh, and [`irq_common`](https://gist.github.com/mark-i-m/361cbcc39769f965b1c419091b9cbf4f#file-machine-rs-L248) is another great example! It has a lot of stuff
- naked
- inlined
- inline assembly with outputs to Rust variables, including a pointer to a Rust-defined struct
- calls a normal Rust function
- returns `!`
- has a `panic` which should never be called.

**Comment by mark-i-m** [comment]
I guess since embedded is a focus in 2018, this will be revived soon.

One thing to note is that this feature is pretty useless to stabilize before inline asm 😛

**Comment by dancrossnyc** [comment]
It would be nice to see some movement on naked functions; I'm using them for e.g. interrupt handlers and the like in kernel-level code. Naked functions are one of the only ways to inject e.g. Rust constants into things like interrupt and syscall handlers. Yes, one can do that via inline assembly in "normal" functions, but there the preamble and epilogue can trash things you'd like to preserve on your stack (if you even have one! Consider the behavior of the `syscall` instruction on x86_64: the CPU switches into kernel mode and start executing whatever the LSTAR MSR points at, but at that point you don't even have a kernel stack!).

Module level assembler doesn't really work here since you can't inject the value of a `const` into the text presented to the "assembler". If I want to use the value of a Rust constant in assembler, I've pretty much got to pass it as an immediate in an inline assembler statement a la 'asm!()'.

We're getting concerned about the dependency on unstable features and ties to the nightly compiler to get access to those unstable features, and what production people might have to say about that (which could potentially scuttle our entire Rust effort). Put another way, it sure would be nice to be able to build a kernel using only stable Rust.

**Comment by Amanieu** [comment]
I think the general consensus is that we are happy to have naked functions, but there is no point in stabilizing them before inline asm since they are useless without it.

**Comment by mark-i-m** [comment]
Lol, that sounds like a challenge to find something you can do with naked functions without assembly :laughing: 

```rust
#[naked]
fn hehehe() -> ! {
  loop {} // No inline asm here!
}
```

----

More seriously, I think naked functions can actually be useful apart from inline-asm. I was thinking about what specifically `naked` means... I think it means the following:

- The "nakedness" of a function is invisible to the caller. The caller must obey the calling convention but doesn't have any other special responsibilities. Among other things, this means that the _caller_ must keep its stack pointer up to date (or at least, it should point to the top of the stack at the time of the naked function call).
- The naked function does not have a function prologue or epilogue. If it wants stack space, it has to make it using inline asm.
- Using stack-allocated locals (e.g. let-bindings) in a `naked` function is unspecified behavior. I suggest we prohibit it for now.
- The naked function promises to obey all calling convention contracts with its caller (e.g. not clobbering callee-saved registers or the caller's stack frame).
- Naked functions can safely make function calls to other functions. Their arguments are pushed to the stack before the call and popped afterwards, just like any normal function call.

In light of this, the following examples seem like they would be valid:

```rust
#[no_mangle]
fn unsafe extern "x86-interrupts" do_pic_irq(irq_stack: IrqStackFrame) {
  sched.timer(&irq_stack);
}
```

```rust
#[no_mangle]
fn unsafe extern "x86-interrupts" do_page_fault(exception_stack: ExceptionStackFrame) {
  CURRENT_PROCESS.mm.handle(&exception_stack);
}
```

**Comment by main--** [comment]
>I think it makes sense to just enforce that all naked fns have to be extern C.

I'm not even sure if this is sufficient. What about arguments? Obviously declaring function parameters is out of the question - you would have to do it in your asm prologue instead. But this opens up the interesting problem that you can no longer have a volatile inline asm block as the first thing in the function: you need to at least declare (and potentially initialize?) the variables that are going to hold your parameters first. Is this guaranteed to be safe? Given that Rust does not have a `register` keyword like C I don't see why the compiler shouldn't be allowed to stack-allocate space for a variable prior to the prologue in this case. To avoid *this*, compilers typically don't allocate stack space in naked functions at all, instead relying on the user-supplied prologue to do this as well. But how would you know exactly how many bytes to reserve? Who specifies that this is always directly measured from esp? I think this is no really about calling conventions but rather about the very specific interaction between a naked function and the compiler-generated code inside of it.

I get the feeling that naked functions are a minefield of UB and footguns that should be avoided. Perhaps custom ABIs / calling conventions could be a better way to solve the same problems?

**Comment by mark-i-m** [comment]
> Obviously declaring function parameters is out of the question - you would have to do it in your asm prologue instead.

I didn’t followed this point. The arguments are not on the callee’s stack frame; they are on the caller’s stack frame. As long as the caller obeys the calling convention, what’s the problem?

> I get the feeling that naked functions are a minefield of UB and footguns that should be avoided. Perhaps custom ABIs / calling conventions could be a better way to solve the same problems?

I think that would lead to a proliferation of custom calling conventions. Moreover fully specifying a calling convention sounds like a pain.

**Comment by main--** [comment]
>I didn’t followed this point. The arguments are not on the callee’s stack frame; they are on the caller’s stack frame. As long as the caller obeys the calling convention, what’s the problem?

Since the function is naked the compiler cannot generate code to access the parameters, or what am I missing?

**Comment by mark-i-m** [comment]
> Since the function is naked the compiler cannot generate code to access the parameters, or what am I missing?

Oh, I see what you are saying now. I was thinking from the other side: the compiler _can_ generate code to pass the parameters, but yes, I think the naked function would then need to write it's own code to access them.

We could enforce that all arguments to the naked function need to be `_: T`...

**Comment by alistair23** [comment]
I have opened an RFC on the argument issue: https://github.com/rust-lang/rfcs/pull/2774

**Comment by mark-i-m** [comment]
Feedback appreciated: https://internals.rust-lang.org/t/idea-naked-functions-2-0/11637/14

**Comment by alistair23** [comment]
Looks good, I still think it is nice to allow arguments. Although the code block can't directly access them they can be accessed by the assembly (via the registers the caller used) and then it's clear what arguments we are expecting from the caller.

I'm thinking here about the entry point to a program loaded by a kernel. The program knows that some values will be in registers A, B and C so by having the args in the function (instead of just comments) readers can see what will be passed in. A, B and C are the ISA specified registers for passing arguments, so they do all line up. The rule should be then that Rust code can't access the parameters.

**Comment by pitust** [comment]
naked functions break with code-model large sometimes: rust decides to put something in rax even before the prolouge which can lead to some... _interesting_ issues.

**Comment by Amanieu** [comment]
@pitust Can you provide an example to reproduce this? This is most likely a bug in LLVM.

**Comment by yvt** [comment]
Naked functions generate a trap instruction at the end of them on some target architectures <https://rust.godbolt.org/z/f7To38> - is this an expected behavior? I'd expect naked functions to have absolutely no compiler-generated surrounding code (prologue, epilogue, and of course the trap instruction).

```rust
#[naked]
pub extern "C" fn i_am_naked() {
    unsafe {
        #[cfg(target_arch = "arm")]
        asm!("bx lr", options(noreturn));

        #[cfg(target_arch = "riscv")]
        asm!("ret", options(noreturn));

        #[cfg(target_arch = "x86_64")]
        asm!("ret", options(noreturn));
    }
}
```

`x86_64`:

```noneexample::i_am_naked:
example::i_am_naked:
        ret
        ud2
```

`armv7a`:

```none
example::i_am_naked:
        bx      lr
        .inst   0xe7ffdefe
```

`thumbv6m`:

```none
example::i_am_naked:
        bx      lr
        .inst.n 0xdefe
```

**Comment by Amanieu** [comment]
This is due to LLVM inserting a trap instruction when lowering the `unreachable` IR instruction which we insert after a `noreturn` asm. It shouldn't hurt since it is inserted after the `asm` rather than before.

**Comment by kotauskas** [comment]
I feel like rustc should specifically avoid passing `unreachable` over to LLVM if it reads it in a naked function and only use it for type-checking and MIR optimization. Trapping `ud2` after an unconditional `ret` doesn't seem sensible especially considering that the guarantee of not generating unexpected assembly in the function is the guarantee that naked functions are supposed to have.

**Comment by npmccallum** [comment]
@kotauskas Under the [Constrained Naked definition](https://github.com/rust-lang/rfcs/pull/2972) the compiler commits to avoid generating instructions **before** the `asm!()` but not afterwards.

**Comment by ahmedcharles** [comment]
> @kotauskas Under the [Constrained Naked definition](https://github.com/rust-lang/rfcs/pull/2972) the compiler commits to avoid generating instructions **before** the `asm!()` but not afterwards.

The RFC doesn't seem to address why this is the case, so simply stating this doesn't really address the feedback.

**Comment by npmccallum** [comment]
@ahmedcharles I was merely offering it as a data point, not an argument.

I actually came across a case today where the additional instructions after the asm block caused problematic output. The case was where a function needed to be exactly page-sized for a specific hardware use. The author used the fill intrinsic in asm to pad the function size at the end. The additional instructions after the asm block made the function incorrectly sized. Based on this case I'm inclined to commit the compiler to insert no additional instructions at all.

**Comment by Amanieu** [comment]
I think this should just be a matter of disabling LLVM's TrapUnreachable option which Rust had previously enabled in an attempt to deal with infinite-looping being considered UB by LLVM.

**Comment by npmccallum** [comment]
> I think this should just be a matter of disabling LLVM's TrapUnreachable option which Rust had previously enabled in an attempt to deal with infinite-looping being considered UB by LLVM.

I think that will disable the `ud2` but not the `ret`, no?

**Comment by roblabla** [comment]
> The case was where a function needed to be exactly page-sized for a specific hardware use.

I feel like that use-case would be better served by `global_asm`, which makes it much easier to reason about what's going on.

**Comment by npmccallum** [comment]
> > The case was where a function needed to be exactly page-sized for a specific hardware use.
> 
> 
> 
> I feel like that use-case would be better served by `global_asm`, which makes it much easier to reason about what's going on. 

At the cost of higher level features such as visibility, namespaces, docs, etc. The last one is the most important. I'd prefer clean documentation about such a strange case versus `global_asm`.

**Comment by fintelia** [comment]
What does it even mean to require no instructions after the end of an asm block? Does it have to be the very last thing in the binary? Outside of specific regions that are defined using assembly, the compiler generally has free reign to place any instructions it wants anywhere in the binary.

**Comment by npmccallum** [comment]
> What does it even mean to require no instructions after the end of an asm block?

It means that size of the emitted code is exactly the same as the instructions/intrinsics in the `asm!` block.

> Does it have to be the very last thing in the binary?

No. But when the size of the output matters, it is usually placed into its own section. For example, the fill technique mentioned above produces a function with exactly 4096 bytes. Via `#[link_section]`, it produces a custom section with exactly 4096 bytes. The loader that is parsing that binary then does something special with that section and expects it to be a certain size. This may or may not be coupled with a custom linker script to choose things like section alignment and program header composition.

> Outside of specific regions that are defined using assembly, the compiler generally has free reign to place any instructions it wants anywhere in the binary.

The compiler has free reign to place any bytes (instructions or otherwise) it wants in the standard sections (`.text`, `.data`, `.rodata`, `.bss`, et cetera) of the binary. It places nothing in non-standard sections unless directed to using `#[link_section]`. A developer who needs to use `#[link_section]` should have an extraordinary amount of control over the output bytes. This is achieved for data types via `#[repr(...)]` and for functions via `#[naked]`.

Relatedly, we should permit `#[align(n)]` (or similar) on functions. [GCC/clang allow this](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-aligned-function-attribute) (for the reasons identified above). Supporting this makes custom linker scripts unnecessary for a number of use cases. However, this feature is out of scope for this issue.

**Comment by npmccallum** [comment]
Here's a full example of the situation. This Rust code:

```rust
#![feature(asm, naked_functions)]

/// This function has a fixed size in a custom section.
// It would be nice if we could do this here: #[align(4096)]
#[naked]
#[link_section = ".custom.section"]
pub extern "C" fn inc(x: usize) -> usize {
    unsafe {
        asm!(
            ".Linit:",
            "inc rdi",
            "mov rax, rdi",
            "ret",
            ".Lfini:",
            ".fill((4096 - (.Lfini - .Linit)))",
            options(noreturn)
        );
    }
}
```

Produces this assembly output:

```assembly
	.text
	.intel_syntax noprefix
	.file	"playground.1dff1134-cgu.0"
	.section	.custom.section,"ax",@progbits
	.globl	_ZN10playground3inc17hc4c6ef74646664cfE # -- Begin function _ZN10playground3inc17hc4c6ef74646664cfE
	.p2align	4, 0x90
	.type	_ZN10playground3inc17hc4c6ef74646664cfE,@function
_ZN10playground3inc17hc4c6ef74646664cfE: # @_ZN10playground3inc17hc4c6ef74646664cfE
	.cfi_startproc
# %bb.0:
	#APP

.Linit:
	inc	rdi
	mov	rax, rdi
	ret
.Lfini:
	.fill	4096-(.Lfini-.Linit), 1, 0x0

	#NO_APP
	ud2
.Lfunc_end0:
	.size	_ZN10playground3inc17hc4c6ef74646664cfE, .Lfunc_end0-_ZN10playground3inc17hc4c6ef74646664cfE
	.cfi_endproc
                                        # -- End function
	.section	".note.GNU-stack","",@progbits
```

Note that the stray `ud2` instruction means that the `.custom.section` section is not `4096` bytes like expected.

**Comment by Amanieu** [comment]
Workaround: `-Z trap-unreachable=no`

This was originally added as a way to avoid UB for infinite loops, but this is no longer needed thanks to recent LLVM changes. We could consider disabling this by default now.

**Comment by kotauskas** [comment]
Well, non-UB infinite loops aren't the only purpose of `trap-unreachable`, are they? It's still generally useful to have the program halt when it reaches a normally unreachable code branch instead of just executing arbitrary code. Extra safety checks like this never hurt anyone, especially if they have essentially zero overhead when not hit. Having to disable this functionality for a whole crate to get the desirable naked function behavior is definitely a suboptimal solution.

**Comment by roblabla** [comment]
If we want naked functions to guarantee that no extra assembly is generated after the asm, it should be documented and tested. Otherwise, programs would end up relying on an implementation detail that may regress at any time. Currently, all the RFCs and docs guarantee is the lack of extra instructions before the inline asm body, nothing about trailing instructions.

**Comment by ranma42** [comment]
IIUC if the asm block is not marked `noreturn`, the compiler will generate the code for whatever follows the block, including a final `ret`, while if it is marked `noreturn` only a `ud2` is generated (and all of the code is omitted).
Would it make sense to have `options(untrapped_noreturn)` that guarantees that the compiler emits nothing?

**Comment by ranma42** [comment]
Also the (admittedly outdated) [example in the RFC](https://github.com/rust-lang/rfcs/blob/master/text/1201-naked-fns.md#example) used an explicit `intrinsics::unreachable()` after the `asm!` block. Maybe we could avoid automatically inserting the `unreachable` IR instruction and let the developers do it if/when they want the additional protection (since they are manually writing the function, including the prologue and epilogue, in unsafe assembly, they can probably be trusted to know better than the compiler).

**Comment by haraldh** [comment]
@ranma42 Both solutions are way better than any unexpected compiler generated code. I personally expect total control, if I use add the `naked` attribute. Even the Summary of the RFC states "epilogue-free".

If I put that function in a link section, and only that function, I don't expect that link section to contain more bytes than my function. If rust cannot guarantee that, then I will have to switch to pure assembler again, which would be sad.

**Comment by fintelia** [comment]
Another option would be to insert the `ud2` instruction only if the function is located in the .text segment. This would potentially catch bugs where execution runs off the end of naked functions, while not creating other issues because the compiler is allowed to place padding instructions between functions in the text segment whenever it wants.

**Comment by Amanieu** [comment]
We don't have direct control over the `ud2` instruction, it is inserted by LLVM because the `TrapUnreachable` option is enabled. Removing the `unreachable` IR instruction won't help: LLVM requires that all basic blocks end with a terminator instruction in the IR.

The only way to eliminate it is to disable `TrapUnreachable` by default.

**Comment by npmccallum** [comment]
I'm [not able to reproduce the issue](https://godbolt.org/z/5P76n53Pb) with `clang`. I presume because `TrapUnreachable` is disabled. But I'm not sure how to test that thesis.

Also, it seems that LLVM [has some similar precedent](https://reviews.llvm.org/D66980).

**Comment by Amanieu** [comment]
I did a bit of searching in LLVM's code and it seems that `TrapUnreachable` is actually required on some targets because the unwinder requires that call return addresses point within the bounds of the caller function. This requires adding an `ud2` (or equivalent) at the end of a function to ensure the byte after a `noreturn` call is still inside the function.

In conclusion I don't think that we can actually guarantee the lack of any instructions in the epilogue with the current implementation in LLVM. Possible solutions:
- Recommend using `global_asm!` instead.
- Reimplement naked functions on top of `global_asm!` internally.

**Comment by roblabla** [comment]
> Reimplement naked functions on top of `global_asm!` internally.

While this sounds simple on the surface, I believe the difficulty comes from identifying the attributes we can attach to a function definition (`export_name`, `link_section`, `cfg`, `target_feature`, etc...) and figuring out how to translate them to whatever is necessary for the `global_asm!`.

**Comment by mark-i-m** [comment]
> Reimplement naked functions on top of global_asm! internally.

I was actually thinking about this the other day. I think this would allow creating custom calling conventions (e.g. for interrupt handlers or custom syscall handling or whatever or accepting args to a naked function), which is something I've wanted for a while...

**Comment by hudson-ayers** [comment]
I will second that the [Tock project](https://github.com/tock/) also uses naked functions extensively, would prefer to use them over `global_asm!()`, and are interested in them being stabilized. Once `asm!()` is stabilized, naked_functions will be one of the last remaining blockers from us moving some of our libraries to a stable compiler.

**Comment by roblabla** [comment]
> I did a bit of searching in LLVM's code and it seems that TrapUnreachable is actually required on some targets because the unwinder requires that call return addresses point within the bounds of the caller function. This requires adding an ud2 (or equivalent) at the end of a function to ensure the byte after a noreturn call is still inside the function.

What if we forced the naked functions to be nounwind? I don't think I've ever seen a naked function that called an unwinding function. And anyways, if it did, I'd expect it to deal with the unwinding itself, through assembler directives (that's rather the point of naked functions - they are supposed to deal with everything).

**Comment by npmccallum** [comment]
@roblabla That may not be a bad suggestion. You might want to start a discussion on the [Constrained Naked Functions RFC](https://github.com/rust-lang/rfcs/pull/2972) since it is in the FCP. It would be difficult to adopt this after we stabilized the feature with the existing requirements.

**Comment by roblabla** [comment]
Ah, that's true, thanks for the heads up!

**Comment by joshtriplett** [comment]
With constrained naked functions now accepted and merged ([tracking issue](https://github.com/rust-lang/rust/issues/90957)), the @rust-lang/lang team meeting today discussed "unconstrained" naked functions, and we're generally in favor of closing this (and un-accepting RFC 1201) in favor of constrained naked functions.

However, we'd like to give time for folks who are following this issue to confirm that constrained naked functions meets their use cases. To give that time, and to confirm the in-meeting consensus:

@rfcbot close

**Comment by npmccallum** [comment]
@joshtriplett I support this direction. However, constrained naked functions is basically a set of lints on top of naked. I suspect the two are roughly merged at this point.

For example, in the constrained naked tracking issue, there is a feature flag. But I don't think anything uses it since all the lints apply on the regular naked feature flag.

We are very close and I wonder if @bstrie would be willing to push this over the line into stable.

**Comment by nikomatsakis** [comment]
@npmccallum can you expand, I don't quite understand what you wrote -- 

> We are very close and I wonder if @bstrie would be willing to push this over the line into stable.

Close to what? How I understand this is this:

* Constrained naked functions (rust-lang/rfcs#2972) means "naked functions whose body is an inline assembly"
* "General" naked functions (this issue, rust-lang/rfcs#1201) means "naked functions whose body is something other than just inline assembly"

It seems like we are no closer than we ever were to having a clear idea what the latter *means* -- i.e., in what cases is it legal, what kind of things can you do or not do in that code. Is that correct? Also, what use cases are there for a naked function whose body is something other than inline assembly? (Not saying there aren't any, I just don't know what they are; it'd be helpful to know what kind of code people would like to write there)

**Comment by npmccallum** [comment]
@nikomatsakis At this point we have two feature flags:

* `naked_functions` - this currently gates the `#[naked]` function attribute.
* `constrained_naked` - this currently gates nothing (unless I've missed something).

The `constrained_naked` flag was created after the Constrained Naked RFC (2972) was accepted. However, the actual functionality described in RFC 2972 was implemented as a set of restricting lints on top of `naked_functions`. All these lints were implemented before RFC 2972 was finally accepted and therefore before the `constrained_naked` feature flag was created.

Today, if you enable `naked_functions` and use the `#[naked]` attribute, the compiler will error or warn (as appropriate) if you violate the definition of constrained naked functions. Therefore `naked_functions` *is* constrained naked functions.

I propose we dump `constrained_naked` and stabilize `naked_functions` since it is now governed by RFC 2972. This implies that we basically need to merge #32408 and #90957 in whatever way is sensible.

**Comment by joshtriplett** [comment]
@npmccallum That sounds accurate, from a practical standpoint. Effectively, it sounds like we should "redefine" `naked_functions` as being the feature flag associated with RFC 2972 naked functions, and stabilize it when we want to stabilize `#[naked]` as defined by RFC 2972.

The point of this FCP close was to acknowledge that we're not planning to add non-asm naked functions as defined by RFC 1201, and that `#[naked]` won't be expanded to how RFC 1201 defined it.

**Comment by npmccallum** [comment]
@joshtriplett I support that overall move. I just want to make sure we don't accidentally drop something important given the slightly odd history of how this feature developed. Overall, I'm flexible. I just want any breakage to be deliberate rather than accidental.

Another option would be to move all the `naked_functions` functionality to be gated by the `constrained_naked` flag. But that strikes me as churn that causes pain for little gain. However, if we prefer to keep the feature process "clean" then I can get on board with that too. But I suspect the "redefinition" strategy is probably the better one.

Please let me know what I should do to make sure this situation gets normalized.

**Comment by joshtriplett** [comment]
@npmccallum I think we just want to confirm this FCP, and then process a minor PR doing the redefinition.

That's separate from stabilization of constrained naked functions, which I think will take 1) a bit more time, since they've only been around for ~2 months and 2) some reports to https://github.com/rust-lang/rust/issues/90957 confirming that people are using them in nightly and they work as expected.

**Comment by npmccallum** [comment]
1. CNFs have been fully implemented for ~6mo and substantially implemented even earlier than that. CNFs also are a restriction in compiler behavior on top of naked functions which have been in nightly for years. There's no new functionality, just new restrictions on old functionality that we can't support.

2. Naked functions (now constrained by CNF lints) are widely used in major projects and prominent library crates. In particular, the Enarx project has been using them for years and, besides adjusting for CNF lints, we've had only one bug. Fixing that bug is what prompted me to write the CNF RFC in the first place.

CNFs should be viewed as the last major remaining polish on a multi-year NF effort rather than a new feature that was just accepted. This is precisely why I've argued that this isn't "reject NFs in favor of CNFs" but rather "CNFs reduce the functionality of NFs to make them stabilizable." One way to articulate that is that CNFs should really be named `min_naked_functions` in the same spirit as `min_const_generics`. It represents the minimum set of naked function features that could be stabilized.

**Comment by joshtriplett** [comment]
@npmccallum Thanks for clarifying! In that case, constrained naked functions just need a PR to stabilize, and a stabilization report (of which your comment here is most of what's needed).

**Comment by nikomatsakis** [comment]
@npmccallum that all sounds fine, with just one thing to double check:

> However, the actual functionality described in RFC 2972 was implemented as a set of restricting lints 

The term "restricting lints" suggests to me that people can still write arbitrary code in naked functions if they "allow" the lint. Is that correct?

**Comment by npmccallum** [comment]
@nikomatsakis No. The CNF RFC defines both error cases (MUST clauses) and warning cases (SHOULD clauses). You can disable the warnings but not the errors. I've been using the term "lints" to cover both the errors and warnings, but I understand how that was potentially misleading.

**Comment by ds84182** [comment]
Rerouting the issue I bought up in #93153 surrounding the constraints introduced by CNF:

Currently, the requirement for a single ASM block is quite limiting, while there exists a workaround for macros producing ASM snippets, it currently falls apart when you need to handle constants, registers or symbols. Other than implementation simplicity, it's not clear why this requirement exists.

This requirement prevents some straightforward patterns that could be implemented via macros.

----

For a concrete example, consider an architecture-independent naked fn for implementing a trampoline to an interrupt handler:

```rust
#[naked]
unsafe extern "C" fn interrupt_entry() {
  begin_interrupt!(); // Prepares interrupt stack if needed. Saves registers to stack.
  c_call!(interrupt_handler); // Calls an extern "C" fn implemented in Rust.
  // Requires the use of a `sym` argument.
  end_interrupt!(); // Restores registers from stack. Returns from the interrupt handler.
}
```

Then these macros can be implemented by each supported architecture, and shared across different interrupt handler implementations. Alternatively, they can be provided as a library of assembly snippets.

----

Another example where this would be handy is macros for coprocessor register accesses. Or macros for `memcpy` and `memset` snippets that don't require a function call (great for scenarios without a stack). Or writing generated metadata, etc.

**Comment by bstrie** [comment]
@ds84182 That seems useful, but it seems like it would be fully backwards-compatible to stabilize naked_functions as-is and only later extend it to allow multiple `asm!` blocks. IOW, whether or not your proposal is accepted, it doesn't impede the consideration of this attribute for stabilization.

**Comment by npmccallum** [comment]
@bstrie Correct. The CNF RFC explicitly states: "This definition may be overly strict. There is certainly some code that would work without this. ... It might also be possible to reasonably ease the constraints over time."

From my perspective, I'd like to get something overly restrictive available for use in stable. Over time we can relax those restrictions or add features. But I don't want to let the perfect be the enemy of the good.

**Comment by mark-i-m** [comment]
One thing to consider: is any of the requirements on CNF not relaxable? For example, is there any future compatibility hazard in the `noreturn` requirement that would, say, change what kind of code is emitted after a naked function or change the state of the stack at different points in the function?

**Comment by npmccallum** [comment]
@mark-i-m I think we'd have to consider that on a feature-by-feature basis.

In your particular example, I don't think it is ever the intent to allow anything except `asm` blocks in a naked function. So, for example, we could relax the requirement to have a single asm block and the last asm block must have `noreturn`. But we can't ever let Rust do the return since returning in Rust almost always emits stack manipulation.

**Comment by bstrie** [comment]
I have submitted a stabilization proposal for the `naked_functions` feature in the tracking issue for RFC 2972: https://github.com/rust-lang/rust/issues/90957#issuecomment-1028297041 . I suggest that we continue further discussion in that thread. Since there has been some confusion regarding the relationship between RFC 1201 and RFC 2972, allow me to clarify: RFC 2972 supersedes RFC 1201 by defining a smaller legal subset of naked functions; it is essentially `min_naked_functions` (see also `min_const_generics`, `min_specialization`, etc.) but rather than using a new feature flag it has simply re-used the original `naked_functions` feature flag. The `naked_functions` feature as implemented in the compiler today conforms to RFC 2972, not RFC 1201. I propose that this tracking issue be closed, to keep discussion centralized.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

I am in favor of NOT doing "extended naked functions" and just doing the constrained version.

**Comment by bstrie** [comment]
I am not on the lang team, but since the FCP has ended I hope it is not presumptuous of me to close this in acknowledgement of the lang team's decision to supersede RFC 1201 with RFC 2972. Allow me to once again **strongly** encourage anyone interested in naked functions to please review the tentative stabilization report for RFC 2972's minimal subset of naked functions ( https://github.com/rust-lang/rust/issues/90957#issuecomment-1028297041 ); that stabilization is not yet in FCP, but it may become so relatively soon, so now is the time to bring up any concerns you may have.
